sanitycheck: Fix incorrect QEMUHandler status reporting.

stephanosio opened the pull request:
The current QEMUHandler implementation in sanitycheck does not check
for the process exit code and reports "PASS" even when either the QEMU
executable cannot be launched or exited immediately due to an error
(e.g. unsupported machine type, missing file, ...).

This commit adds QEMU process exit code check and error reporting when
the exit code is a non-zero value.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
sanitycheck: Fix incorrect QEMUHandler status reporting.

The current QEMUHandler implementation in sanitycheck does not check
for the process exit code and reports "PASS" even when either the QEMU
executable cannot be launched or exited immediately due to an error
(e.g. unsupported machine type, missing file, ...).

This commit adds QEMU process exit code check and error reporting when
the exit code is a non-zero value.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

marc-hb commented:
I'm not very familiar with this part of the code but I see a mismatch between the commit message and the code change: the commit message says "add error checking..." but doesn't mention the most visible change: moving code from one class to a completely different one. Explaining the reason for this distant move might help with the review?

PS: I did look at the actual commit message and not just at the description Duplicated & Diverged by github.

stephanosio commented:
> Explaining the reason for this distant move might help with the review?

I moved the code under `if instance.handler.type_str == "qemu":` to inside `QEMUHandler.handle` in order to align with what other handlers are doing (note that if `instance.handler.type_str` is `qemu`, the `instance.handler` must be `QEMUHandler`).

https://github.com/zephyrproject-rtos/zephyr/blob/0297756342fb60e6a64813aac046f4d241cbd369/scripts/sanitycheck#L1937-L1938

For example, the `handle` method of both `BinaryHandler` and `DeviceHandler` execute the "runner", wait for it to finish and set the result; there is no reason for `QEMUHandler` to be an exception.

`handle` method implementation for `BinaryHandler`:
https://github.com/zephyrproject-rtos/zephyr/blob/0297756342fb60e6a64813aac046f4d241cbd369/scripts/sanitycheck#L559-L635

Note that `subprocess.Popen`, `proc.wait()`, `self.set_state(...)` and `self.instance.reason =` are all inside the `handle` method.

marc-hb commented:
Thanks, looks like some of explanation above belong to the commit message(s) 
https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes

stephanosio commented:
> Thanks, looks like some of explanation above belong to the commit message(s)

I will split the exit code check and the move into two separate commits (the move was actually an afterthought).

stephanosio committed:
sanitycheck: Relocate QEMU-specific code to QEMUHandler.

This commit relocates the QEMU-specific code that currently resides in
ProjectBuilder.run to QEMUHandler.handle, in order to align with what
other handlers are doing.

For more details, refer to the PR #20573.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>
